sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -175,6 +175,64 @@ object GpuCast extends Arm {
       cv.stringReplaceWithBackrefs(rule.search, rule.replace)
     })
   }
+
+  def sanitizeStringToFloat(input: ColumnVector): ColumnVector = {
+
+    // This regex gets applied after the transformation to normalize use of Inf and is
+    // just strict enough to filter out known edge cases that would result in incorrect
+    // values. We further filter out invalid values using the cuDF isFloat method.
+    val VALID_FLOAT_REGEX =
+      "^" +                         // start of line
+      "[+\\-]?" +                   // optional + or - at start of string
+      "(" +
+        "(" +
+          "(" +
+            "([0-9]+)|" +           // digits, OR
+            "([0-9]*\\.[0-9]+)|" +  // decimal with optional leading and mandatory trailing, OR
+            "([0-9]+\\.[0-9]*)" +   // decimal with mandatory leading and optional trailing
+          ")" +
+          "([eE][+\\-]?[0-9]+)?" +  // exponent
+          "[fFdD]?" +               // floating-point designator
+        ")" +
+        "|Inf" +                    // Infinity
+        "|[nN][aA][nN]" +           // NaN
+      ")" +
+      "$"                           // end of line
+
+    withResource(input.lstrip()) { stripped =>
+      withResource(GpuScalar.from(null, DataTypes.StringType)) { nullString =>
+        // filter out strings containing breaking whitespace
+        val withoutWhitespace = withResource(ColumnVector.fromStrings("\r", "\n")) {
+            verticalWhitespace =>
+          withResource(stripped.contains(verticalWhitespace)) {
+            _.ifElse(nullString, stripped)
+          }
+        }
+        // replace all possible versions of infinity with Inf
+        val inf = withResource(withoutWhitespace) { _ =>
+          withoutWhitespace.stringReplaceWithBackrefs(
+            "([iI][nN][fF])([iI][nN][iI][tT][yY])?", "Inf")
+        }
+        // replace +Inf with Inf because cuDF only supports "Inf" and "-Inf"
+        val infWithoutPlus = withResource(inf) { _ =>
+          withResource(GpuScalar.from("+Inf", DataTypes.StringType)) { search =>
+            withResource(GpuScalar.from("Inf", DataTypes.StringType)) { replace =>
+              inf.stringReplace(search, replace)
+            }
+          }
+        }
+        // filter out any strings that are not valid floating point numbers according
+        // to the regex pattern
+        val floatOrNull = withResource(infWithoutPlus.matchesRe(VALID_FLOAT_REGEX)) {
+            _.ifElse(infWithoutPlus, nullString)
+        }
+        // strip floating-point designator 'f' or 'd' but don't strip the 'f' from 'Inf'
+        withResource(floatOrNull) {
+          _.stringReplaceWithBackrefs("(Inf|-Inf|[nN][aA][nN]|[^InNaAfFdD]*)([fFdD]?)", "\\1")
```

Comment:
I'm not sure why this regex needs to be so complicated.  We've already filtered the strings to things that have to look like floats, and as the comment above says, we want to strip trailing `f` or `d` as long as it's not part of `Inf`.  Looking at the `VALID_FLOAT_REGEX` definition above, I don't see the harm in making this a very straightforward regex of essentially "drop trailing f's and d's as long as they aren't preceded by an n".
```suggestion
          _.stringReplaceWithBackrefs("([^n])[fFdD]?$", "\\1")
```

Reply:
I did try variations on this and couldn't find anything that worked. The challenge is that we need to retain the `n` in the capture group so that we retain the `Inf` to pass to cuDF. Perhaps I am just missing some regex trick here though.

Reply:
> The challenge is that we need to retain the n in the capture group

But the regex example I showed above shouldn't hit at all on `Inf` because it's looking for anything _except_ an `n` character to capture before it finds a case-insensitive `f` or `d`.  And what it does capture it emits, so I'm totally confused how this could be doing anything at all on `Inf` or how it could eat an `n` character which should alone disqualify the expression from matching.

Reply:
Maybe try dropping the `?` which I think is unnecessary.  We want to strip any `f` or `d` character at the end of the string if it's not preceded by `n`.  If there's no `f` or `d` at the end then we don't want it to match at all and emit the original string.
```
stringReplaceWithBackrefs("([^n])[fFdD]$", "\\1")
```

Reply:
Thanks. That works perfectly. I'm confused now why I couldn't get this to work yesterday.